tests/providers/microsoft/psrp/hooks/test_psrp.py

```diff
@@ -29,43 +30,67 @@
 CONNECTION_ID = "conn_id"
 
 
-class TestPSRPHook(unittest.TestCase):
-    @patch(
-        f"{PSRPHook.__module__}.{PSRPHook.__name__}.get_connection",
-        return_value=Connection(
-            login='username',
-            password='password',
-            host='remote_host',
-        ),
-    )
-    @patch(f"{PSRPHook.__module__}.WSMan")
-    @patch(f"{PSRPHook.__module__}.PowerShell")
-    @patch(f"{PSRPHook.__module__}.RunspacePool")
-    @patch("logging.Logger.info")
-    def test_invoke_powershell(self, log_info, runspace_pool, powershell, ws_man, get_connection):
-        with PSRPHook(CONNECTION_ID) as hook:
-            ps = powershell.return_value = MagicMock()
-            ps.state = PSInvocationState.RUNNING
-            ps.output = []
-            ps.streams.debug = []
-            ps.streams.information = []
-            ps.streams.error = []
+def mock_powershell():
```

Comment:
FWIW you can subclass `Mock` (or `MagicMock`. This could make this a bit more readable?

Reply:
I could change it to something like:
```python
class MockPowerShellFactory(MagicMock):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        ps = self.return_value = MagicMock()
        ps.state = PSInvocationState.NOT_STARTED

        def poll_invoke():
            ps.output.append("<output>")
            ps.streams.debug.append(MagicMock(spec=InformationRecord, message_data="<message>"))
            ps.state = PSInvocationState.COMPLETED

        def begin_invoke():
            ps.state = PSInvocationState.RUNNING
            ps.output = []
            ps.streams.debug = []
            ps.streams.information = []
            ps.streams.error = []

        def end_invoke():
            while ps.state == PSInvocationState.RUNNING:
                poll_invoke()
            ps.streams.error = []

        ps.poll_invoke.side_effect = poll_invoke
        ps.begin_invoke.side_effect = begin_invoke
        ps.end_invoke.side_effect = end_invoke
```

Reply:
I have tried to improve this now in `4ed556aff2b96f1aa35ace02dab669613ebc77a3`. Seems clearer now at least.